templates/nextflow-s3-template.yaml

```diff
@@ -0,0 +1,144 @@
+  WriteAccess:
+    Type: CommaDelimitedList
+    # Default: "[]"
+    # Description: Grant write access to accounts, groups, and users.
+    ConstraintDescription: List of ARNs (i.e. ["arn:aws:iam::011223344556:user/jsmith", "arn:aws:iam::544332211006:user/rjones"]
+
+Conditions:
+  HasBucketName: !Not [!Equals [!Ref BucketName, ""]]
+
+# Resources
+Resources:
+  Bucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: !If [HasBucketName, !Ref BucketName, !Ref 'AWS::NoValue']
+      AccessControl: Private
+      # VersioningConfiguration:
+        # SType: String
+        # Description: Enabled to enable bucket versioning, default is Suspended
+        # AllowedValues:
+        #   - Enabled
+        #   - Suspended
+        # Default: Suspended
+      # BucketEncryption:
+      #   ServerSideEncryptionConfiguration:
+      #     - ServerSideEncryptionByDefault:
+      #         SSEAlgorithm: AES256
+
+  # Policy
+  BucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref Bucket  
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+
+          # - Sid: DenyAllBucketAccess
+          #   # denies access to everyone
+          #   Effect: Deny
+          #   Principal:
+          #     AWS: "*"
+          #   Action:
+          #     - "s3:*"
+          #   Condition:
+          #     StringNotLike:
+          #       "aws:PrincipalAccount":
+          #         - arn:aws:iam::035458030717:root
+          #         - arn:aws:iam::035458030717:role/aws-reserved/sso.amazonaws.com/AWSReservedSSO_Administrator_580e9f32ac55c4e7
+          #   Resource:
+          #     - !Sub arn:aws:s3:::${Bucket}/*
+          #     - !Sub arn:aws:s3:::${Bucket}
+
+          - Sid: BucketLocationAccess
+            # gives grantees access to the bucket
+            Effect: Allow
+            Principal:
+              AWS: !Join
+                - ","
+                - - Fn::ImportValue: !Sub "${AWS::Region}-nextflow-forge-service-user-ServiceUserArn"
+                  - !Join [",", !Ref BucketAccess]
+            Action:
+              - "s3:ListBucket"
+              - "s3:ListBucket*"
+              - "s3:GetBucketLocation"
+            Resource:
+              - !Sub arn:aws:s3:::${Bucket}/*
+              - !Sub arn:aws:s3:::${Bucket}
+
+          - Sid: ReadObjectAccess
+            # give grantees read access to objects
+            Effect: Allow
+            Principal:
+              AWS: !Join
+                - ","
+                - - Fn::ImportValue: !Sub "${AWS::Region}-nextflow-forge-service-user-ServiceUserArn"
+                  - !Join [",", !Ref ReadAccess]
+            Action:
+              - "s3:GetObject"
+              - "s3:GetObjectAcl"
+              - "s3:AbortMultipartUpload"
+              - "s3:ListMultipartUploadParts"
+            Resource:
+              - !Sub arn:aws:s3:::${Bucket}/*
+              - !Sub arn:aws:s3:::${Bucket}
+
+          - Sid: WriteAccess
```

Comment:
does a principal need to be added to all three (BucketAccess, ReadAccess, and WriteAccess) to have write access to objects in this bucket? or does the principal only need to be in WriteAccess group to have write access to the bucket and objects?

Reply:
Just edited to allow users in WriteAccess to have permissions in BucketAccess and ReadAccess! So the principal only needs to be in the WriteAccess group to have access to the bucket and objects